### ISSUE TITLE
Possible issue in RDFProtonDescriptor_G3R?

### ISSUE THREAD

**Issue by Danissss** [cdk/cdk#620]
Hi,
For RDFProtonDescriptor_G3R (Atomic Descriptor), my colleague found one issue about inconsistent return values 
```
calculating==============:=> H
// 13 (g3r_1 -> g3r_13)
[g3r_1, g3r_2, g3r_3, g3r_4, g3r_5, g3r_6, g3r_7, g3r_8, g3r_9, g3r_10, g3r_11, g3r_12, g3r_13] 
// 14 return values
1.999947761697666,1.6993980019913368,1.0339422629849824,0.4504247076719672,0.1404988780501808,0.03137966047684351,0.005018204523279824,5.746100374003407E-4,4.711105086599298E-5,2.765649302901093E-6,1.1625085584520994E-7,3.498806641542736E-9,7.539949876924451E-11,1.163433885460145E-12
```

Another one 
```
calculating==============:=> H
[g3r_1, g3r_2, g3r_3, g3r_4, g3r_5, g3r_6, g3r_7, g3r_8, g3r_9, g3r_10, g3r_11, g3r_12, g3r_13]
// 13 return NaN
NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
```

Here is the compound we tested:
```
2244
  -OEChem-01102018403D

 21 21  0     0  0  0  0  0  0999 V2000
    1.2333    0.5540    0.7792 O   0  0  0  0  0  0  0  0  0  0  0  0
   -0.6952   -2.7148   -0.7502 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.7958   -2.1843    0.8685 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.7813    0.8105   -1.4821 O   0  0  0  0  0  0  0  0  0  0  0  0
   -0.0857    0.6088    0.4403 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7927   -0.5515    0.1244 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7288    1.8464    0.4133 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1426   -0.4741   -0.2184 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.0787    1.9238    0.0706 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.7855    0.7636   -0.2453 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.1409   -1.8536    0.1477 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.1094    0.6715   -0.3113 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.5305    0.5996    0.1635 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.1851    2.7545    0.6593 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.7247   -1.3605   -0.4564 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.5797    2.8872    0.0506 H   0  0  0  0  0  0  0  0  0  0  0  0
   -3.8374    0.8238   -0.5090 H   0  0  0  0  0  0  0  0  0  0  0  0
    3.7290    1.4184    0.8593 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.2045    0.6969   -0.6924 H   0  0  0  0  0  0  0  0  0  0  0  0
    3.7105   -0.3659    0.6426 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2555   -3.5916   -0.7337 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  5  1  0  0  0  0
  1 12  1  0  0  0  0
  2 11  1  0  0  0  0
  2 21  1  0  0  0  0
  3 11  2  0  0  0  0
  4 12  2  0  0  0  0
  5  6  1  0  0  0  0
  5  7  2  0  0  0  0
  6  8  2  0  0  0  0
  6 11  1  0  0  0  0
  7  9  1  0  0  0  0
  7 14  1  0  0  0  0
  8 10  1  0  0  0  0
  8 15  1  0  0  0  0
  9 10  2  0  0  0  0
  9 16  1  0  0  0  0
 10 17  1  0  0  0  0
 12 13  1  0  0  0  0
 13 18  1  0  0  0  0
 13 19  1  0  0  0  0
 13 20  1  0  0  0  0
M  END
```

**Comment by johnmay** [comment]
Could you give some more context, did you call these one after the other.

**Comment by johnmay** [comment]
...with the same molecule.

**Comment by Danissss** [comment]
Hi, here how I generated atomic descriptors:
```
/**
	 * calculate all descriptor for one given atom
	 * @param mol
	 * @param atom
	 * @param descriptor
	 * @return
	 */
	public static ArrayList<Double> computeDescriptorsAtomic(IAtomContainer mol, IAtom atom, List<IDescriptor> descriptor) {
		ArrayList<Double> vv = new ArrayList<Double>();
		for(int i = 0; i < descriptor.size(); i++) {
			try {
				
				IDescriptorResult res = ((IAtomicDescriptor) descriptor.get(i)).calculate(atom, mol).getValue();
				
				if (res instanceof IntegerResult) {
					vv.add((double) ((IntegerResult) res).intValue());
					
				} else if (res instanceof DoubleResult) {
					vv.add(((DoubleResult) res).doubleValue());
				} else if (res instanceof DoubleArrayResult) {
//					for(int d = 0; d < ((DoubleArrayResult) res).length(); d++) {
//						vv.add(((DoubleArrayResult) res).get(d));
//					}
					
					// did this because of https://github.com/cdk/cdk/issues/620
					String[] descriptorName = ((IAtomicDescriptor) descriptor.get(i)).getDescriptorNames();
					for(int d = 0; d < descriptorName.length; d++) {
						vv.add(((DoubleArrayResult) res).get(d));
					}

					
				} else if (res instanceof IntegerArrayResult) {
					for(int d = 0; d < ((IntegerArrayResult) res).length(); d++) {
						vv.add(Double.valueOf(((IntegerArrayResult) res).get(d)));
					}
				} else if (res instanceof BooleanResult) {
					// true = 1.0; false = 0.0;
					if(((BooleanResult) res).booleanValue() == true) {
						vv.add(1.0);
					}else {
						vv.add(0.0);
					}
				}
				
				else {
					System.out.println("Error");
					throw new IllegalStateException(
							"Unknown idescriptor result value for '" + descriptor + "' : " + res.getClass());
				}
			} catch (Throwable e) {
				System.out.println("Could not compute cdk feature " + descriptor);
				System.err.println("Could not compute cdk feature " + descriptor);
				vv.add(0.0);
			}
		}
		
		
		
		return vv;
	}
```

**Comment by johnmay** [comment]
Let me rephrase, why do you get non-null:

```
calculating==============:=> H
// 13 (g3r_1 -> g3r_13)
[g3r_1, g3r_2, g3r_3, g3r_4, g3r_5, g3r_6, g3r_7, g3r_8, g3r_9, g3r_10, g3r_11, g3r_12, g3r_13] 
// 14 return values
1.999947761697666,1.6993980019913368,1.0339422629849824,0.4504247076719672,0.1404988780501808,0.03137966047684351,0.005018204523279824,5.746100374003407E-4,4.711105086599298E-5,2.765649302901093E-6,1.1625085584520994E-7,3.498806641542736E-9,7.539949876924451E-11,1.163433885460145E-12
```

then 

```
calculating==============:=> H
[g3r_1, g3r_2, g3r_3, g3r_4, g3r_5, g3r_6, g3r_7, g3r_8, g3r_9, g3r_10, g3r_11, g3r_12, g3r_13]
// 13 return NaN
NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
```

was this for the same molecule?

**Comment by Danissss** [comment]
Sorry for late reply. Yes, same molecule but these hydrogen atoms are in different position.

**Comment by johnmay** [comment]
Okay looking through the code, obviously any non-hydrogens get the NaN values since this is only for protons.
However it's more subtle than that and this method only applies to hydrogens attached to "cycloexane" rings. More precisely here is the original description in the paper RDFProtonDescriptor_G3R is:

<img width="334" alt="Screenshot 2022-02-06 at 16 54 07" src="https://user-images.githubusercontent.com/983232/152691782-d3eea1be-e848-454d-8cb6-901512e06ecc.png">

To me this implies you need to use or combine the RDFProtonDescriptor in some way (maybe check the original paper - I only skimmed it). Testing the other ``RDFProtonDescriptor_*`` doesn't seem to be much use but my understanding is these were fed into a model in the paper. It is odd you 0,0,0,0 for the methane but since the G3R seems to be doing what it's meant  to I am going to close this issue. Sorry I can be more help but hopefully given some explanation.

```
G3R       14 H 1.9996761593010985,1.7097759081772612,1.0467511614891674,0.45885180382464263,0.14402111501125947,0.03236716249874062,0.005208441866419566,6.001167489713332E-4,4.95094849424195E-5,2.9245954663133524E-6,1.2369949602877895E-7,3.746233057022608E-9,8.123559898071344E-11
GDR       14 H 0.42518738066312295,0.5388018747457387,0.6141963359287443,0.6296883771256263,0.5807165456083709,0.48205719267146574,0.360593596550044
GHR       14 H -0.004248588268401291,-2.3890733553736185E-4,9.872203384556557E-4,0.01097093818209613,0.03825873750608371,0.05934846756395838,0.07000571192365392,0.04298047165991202,0.008847749306256328,7.376187501272448E-4,0.0028216282371549906,0.012178062750348412,0.013427191296845135,-0.01140852872782842,-0.04262028708476227
GHR_topol 14 H -0.004248789681544038,-2.6528711357789347E-4,-4.18348956866855E-6,7.139521868700133E-5,0.0019250007315369275,0.015598950466001307,0.03800436626487112,0.027838464182408,0.006131304593193769,4.4458719764923463E-4,0.0014382552734818856,0.015941712085357357,0.053960179237229726,0.05952342320071283,0.03157917627308625
GSR       14 H 0.5427614772413394,0.7030870823754738,0.8277521191042343,0.887269611257614,0.867792729657907,0.776128892985477,0.6358751238209304

G3R       15 H 1.9934382685096312,1.7677249377439583,1.122409464052428,0.5102848623762181,0.16611114974954905,0.03871770595495161,0.006461676091607676,7.721567723114003E-4,6.606790254386018E-5,4.047625171341387E-6,1.7755577016542936E-7,5.576915019345126E-9,1.2542323465682913E-10
GDR       15 H 0.22480374239528217,0.2926414524444089,0.3597397328296489,0.423304767439789,0.47958746573873695,0.5208201872299913,0.5355723713300949
GHR       15 H -0.006507213129680704,-4.594715363268573E-4,-9.589299422321463E-4,-0.0025711707071457124,0.013902359775090916,0.059897070268693606,0.14772020320268456,0.29110450612144984,0.2579877455530054,0.0754875460757132,0.011296684473721235,0.026340515172080827,0.022005053094397816,-0.11411109542361801,-0.2829875733127233
GHR_topol 15 H -0.0065064784542550515,-4.067984579208338E-4,-7.316151233277284E-6,2.9543652777824455E-5,7.955679575783582E-4,0.00643173255284429,0.01563331213891942,0.011424742183950877,0.0025105433552395622,2.0408784660563744E-4,0.0014245161911454762,0.016004447517119307,0.05925869162629746,0.12607267618430132,0.2824866435280576
GSR       15 H 0.5308390312660461,0.7087358162529739,0.865392081871494,0.9671657787383344,0.9899490223571565,0.9280697251320904,0.796335167731227

G3R       16 H 1.999994390074378,1.6946720434040894,1.0281755801265544,0.446656477792073,0.13893277847053442,0.030942866069385604,0.004934476513682579,5.634383026952754E-4,4.606556266743283E-5,2.6966906262720768E-6,1.1303439211614304E-7,3.392460602801813E-9,7.290272380305456E-11
GDR       16 H 0.47525474058746253,0.5670599260387397,0.6079928375909764,0.5877235662396832,0.5143936332487211,0.40972114345958566,0.2986792081803279
GHR       16 H -0.008280182147645298,-6.224336533255537E-4,-0.003311703769835809,-0.030286763766550796,-0.07820640327377675,-0.018681309656152188,0.10547012082301213,0.08621722093474934,0.018705420854563337,0.0022838194476112517,0.0126098541513434,0.044142051054801644,0.04965538784926306,0.04331759004736179,0.0759913480408067
GHR_topol 16 H -0.008279130292177193,-5.149199919918622E-4,-1.1527022758657726E-5,-1.691946850106572E-4,-0.004530579073949571,-0.03648667126763282,-0.08834650204378641,-0.06431600205965712,-0.014076782092026207,-8.470043774494672E-4,0.002916325260979645,0.03267251318381033,0.11114344402515608,0.12880808884152634,0.08303482464490168
GSR       16 H 0.5263579716843921,0.7002660817991296,0.8529547729292277,0.9528255972965277,0.9774936741522776,0.9214088511529058,0.7975195695962364

G3R       17 H 1.999947761697666,1.6993980019913368,1.0339422629849824,0.4504247076719671,0.1404988780501807,0.03137966047684346,0.0050182045232798145,5.746100374003408E-4,4.711105086599282E-5,2.765649302901093E-6,1.1625085584520912E-7,3.498806641542717E-9,7.539949876924344E-11
GDR       17 H 0.18399084449050718,0.25801965136733485,0.34041396857687906,0.42408829851131064,0.4975952890714358,0.5456121598040924,0.5533838650681344
GHR       17 H -0.008513625044224062,-6.539759577455804E-4,-0.0038287503097242746,-0.03499775428245859,-0.09027152127176089,-0.025452732852136138,0.10224853670229792,0.07944357938691696,0.016202634536633106,0.0019500934496345338,0.011259759807943076,0.039498871358471264,0.04571076751306125,0.04980050743632726,0.09618995587588082
GHR_topol 17 H -0.008512408517076428,-5.296235098758493E-4,-1.2096907705377756E-5,-1.9529434827021203E-4,-0.005233484243968127,-0.04217846775753739,-0.10220341491288719,-0.07445857648077318,-0.016308796464063583,-9.960930408755757E-4,0.002864394069116696,0.032157854069536,0.10933339050776256,0.12528133034188638,0.07732999593929707
GSR       17 H 0.5213861100635614,0.7083374932700014,0.8742068453102573,0.9835706719174115,1.0119334441099495,0.9540645290682443,0.8249117703452112

G3R       18 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GDR       18 H 0.0,0.0,0.0,0.0,0.0,0.0,0.0
GHR       18 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GHR_topol 18 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GSR       18 H NaN,NaN,NaN,NaN,NaN,NaN,NaN

G3R       19 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GDR       19 H 0.0,0.0,0.0,0.0,0.0,0.0,0.0
GHR       19 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GHR_topol 19 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GSR       19 H NaN,NaN,NaN,NaN,NaN,NaN,NaN

G3R       20 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GDR       20 H 0.0,0.0,0.0,0.0,0.0,0.0,0.0
GHR       20 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GHR_topol 20 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GSR       20 H NaN,NaN,NaN,NaN,NaN,NaN,NaN

G3R       21 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GDR       21 H 0.0,0.0,0.0,0.0,0.0,0.0,0.0
GHR       21 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GHR_topol 21 H NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
GSR       21 H NaN,NaN,NaN,NaN,NaN,NaN,NaN
```

**Comment by johnmay** [comment]
I have added a comment to the JavaDoc to help others in future.
